### PR TITLE
chore(atomic): fix initialization issues when combining Lit interfaces with Stencil components

### PR DESCRIPTION
**Pull request by fpbrault**
This PR fixes initialization issues when combining LIt interfaces and Stencil components.
https://coveord.atlassian.net/browse/KIT-4050

### CODE REVIEW

File: packages/atomic-angular/scripts/build-lit.mjs [hunk 1]
```diff
@@ -1,68 +1,113 @@
...
+}
+
+function processNonLitDeclaration(declaration) {
+  const defineCustomElementFn = `defineCustomElement${declaration.name}`;
+
+  const regex = new RegExp(
+    `@ProxyCmp\\(\\{([^}]*)\\}\\)\\s*\\n@Component\\(\\{([^}]*)\\}\\)\\s*\\nexport class\\s+${declaration.name}\\b`,
+    'g'
+  );
+
```
Comment (fpbrault) [inline packages/atomic-angular/scripts/build-lit.mjs:69]: Since we need a different defineCustomElement behaviour for Lit and Stencil components, we can't use the global defineCustomElements function from the loader.

For Lit components, we need to add a new component declaration that will use a custom defineCustomElement function. For Stencil components, we want to add an import to the defineCustomElement function specific to that component provided by stencil and use it as the defineCustomElementsFn in the @ProxyCmp decorator of the component's declaration.

File: packages/atomic/scripts/cem-plugin.mjs [hunk 2]
```diff
@@ -14,7 +14,14 @@ const moduleLinkPhase = ({moduleDoc}) => {
...
       keptAttributes.length > 0 ? keptAttributes : undefined;
-    declaration.members = [];
+
+    if (!declaration.members) {
+      continue;
+    }
+
+    declaration.members = declaration.members.filter(
+      (member) => member?.privacy !== 'private' && member.kind === 'method'
+    );
```
Comment (fpbrault) [inline packages/atomic/scripts/cem-plugin.mjs:24]: Needed so that the component declarations in atomic-angular for Lit components include their methods (Stencil only includes methods marked with the `@Method` decorator).

File: packages/atomic/src/components/commerce/atomic-product/atomic-product.tsx [hunk 3]
```diff
@@ -232,5 +232,6 @@ export class AtomicProduct {
 
       this.executedRenderingFunctionOnce = true;
     }
+    this.host.classList.add('hydrated');
```
Comment (fpbrault) [inline packages/atomic/src/components/commerce/atomic-product/atomic-product.tsx:235]: Needed for angular

File: packages/atomic/src/components/commerce/product-template/atomic-product-template.tsx [hunk 4]
```diff
@@ -26,7 +26,7 @@ import {
   shadow: true,
 })
 export class AtomicProductTemplate {
-  private productTemplateCommon: ProductTemplateCommon;
+  private productTemplateCommon!: ProductTemplateCommon;
```
Comment (fpbrault) [inline packages/atomic/src/components/commerce/product-template/atomic-product-template.tsx:29]: Needed for angular

File: packages/atomic/src/components/commerce/product-template/atomic-product-template.tsx [hunk 5]
```diff
@@ -60,7 +60,9 @@ export class AtomicProductTemplate {
     string[]
   > = {};
 
-  constructor() {
+  constructor() {}
+
+  connectedCallback() {
```
Comment (fpbrault) [inline packages/atomic/src/components/commerce/product-template/atomic-product-template.tsx:65]: Needed for angular, but it didn't make sense to have this code happen before the component is connected to DOM

File: packages/atomic/src/components/search/atomic-external/atomic-external.tsx [hunk 6]
```diff
@@ -36,11 +38,24 @@ export class AtomicExternal {
...
    * Represents the bound interface for the AtomicExternal component.
    */
   @Prop({mutable: true}) boundInterface?: AtomicInterface;
 
+  connectedCallback() {
+    if (isParentReady(this.#interface)) {
+      markParentAsReady(this.host);
+    }
+  }
+
```
Comment (fpbrault) [inline packages/atomic/src/components/search/atomic-external/atomic-external.tsx:58]: Since atomic-external is an initializableElement but does not actually get initialized itself, we need to mark it as ready when its bound interface is ready.

File: packages/atomic/src/decorators/watch.ts [hunk 7]
```diff
@@ -47,7 +47,7 @@ export function watch<
   Prop extends PublicProperties<Component>,
 >(propName: Prop, options?: WatchOptions) {
   const resolvedOptions: Required<WatchOptions> = {
-    waitUntilFirstUpdate: false,
+    waitUntilFirstUpdate: true,
```
Comment (fpbrault) [inline packages/atomic/src/decorators/watch.ts:50]: Needed for angular

File: packages/atomic/src/lit-components.d.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  let HTMLAtomicCommerceInterfaceElement: {
+    prototype: HTMLAtomicCommerceInterfaceElement;
+    new (): HTMLAtomicCommerceInterfaceElement;
+  };
+  interface HTMLElementTagNameMap {
+    'atomic-commerce-interface': HTMLAtomicCommerceInterfaceElement;
+  }
+ */
+}
+export {};
```
Comment (fpbrault) [inline packages/atomic/src/lit-components.d.ts:25]: We can't have the HTMLElementTagNameMap declarations in the components themselves, as it causes a type conflict when referencing them in atomic-angular. Additionally the HTMLAtomic*Element interfaces in global and Atomic* interfaces in the Components Namespace are needed for the angular sample to build when it references them.

File: packages/atomic/src/utils/initialization-utils.tsx [hunk 9]
```diff
@@ -39,10 +40,20 @@ export function initializeBindings<
...
+
+    if (!parent) {
       reject(new MissingInterfaceParentError(element.nodeName.toLowerCase()));
+      return;
+    }
+
+    if (isParentReady(parent)) {
+      element.dispatchEvent(event);
+    } else {
+      queueEventForParent(parent, event as InitializeEvent, element);
```
Comment (fpbrault) [inline packages/atomic/src/utils/initialization-utils.tsx]: Instead of dispatching the event right away, components now check if their parent initializableElement is initialized first. If the parent is not ready, the event gets added to an event queue specific to that parent. Once the parent marks itself ready, it will dispatch all the queued event associated to itself.

File: packages/atomic/src/lit-components.d.ts (hunk 8 above)
Comment (louis-bompart) [reply packages/atomic/src/lit-components.d.ts:25]: mh... odd... let's split this issue in another ticket

Comment (fpbrault) [reply packages/atomic/src/lit-components.d.ts:25]: This would need to be solved before we merge our lit components.

File: packages/atomic-angular/scripts/build-lit.mjs (hunk 1 above)
Comment (fpbrault) [reply packages/atomic-angular/scripts/build-lit.mjs:69]: I did but some declarations include some extra things in their interface, and have a corresponding proxyOutputs call. It seemed easier to just add the defineCustomElementsFn to the existing declarations instead of handling that corner case. We can always revisit this later

File: packages/atomic/dev/tests/smoke.spec.ts [hunk 10]
```diff
@@ -80,17 +80,7 @@ test.describe('style encapsulation', () => {
...
-      name: 'Search field with suggestions',
-    });
-    await page
-      .locator('atomic-query-summary')
-      .getByText(/Results 1-10 of \d*/)
-      .first()
-      .waitFor();
-    await searchBox.fill('connect.coveo.com');
-    await searchBox.press('Enter');
+    await page.goto('http://localhost:3333/themingTests.html#q=connect');
```
Comment (fpbrault) [inline packages/atomic/dev/tests/smoke.spec.ts:83]: Much simpler this way

File: packages/atomic/scripts/build.mjs [hunk 11]
```diff
@@ -25,6 +26,7 @@ const isCDN = process.env.DEPLOYMENT_ENVIRONMENT === 'CDN';
 const transformers = [
   svgTransformer,
   pathTransformer,
+  versionTransformer,
```
Comment (fpbrault) [inline packages/atomic/scripts/build.mjs:29]: This is for the CoveoAtomic global to work with Lit components

File: packages/atomic/src/components/commerce/atomic-commerce-interface/stencil-atomic-commerce-interface.tsx [hunk 12]
```diff
@@ -208,8 +208,8 @@ export class AtomicCommerceInterface
       return;
     }
 
-    this.context.setLanguage(this.language);
-    this.commonInterfaceHelper.onLanguageChange();
+    //this.context.setLanguage(this.language);
```
Comment (louis-bompart) [inline packages/atomic/src/components/commerce/atomic-commerce-interface/stencil-atomic-commerce-interface.tsx:211]: ?

Comment (fpbrault) [reply packages/atomic/src/components/commerce/atomic-commerce-interface/stencil-atomic-commerce-interface.tsx:211]: Fixed it in my other branch forgot to update it here..!

**Review by dmgauthier — APPROVED** [review]
Not sure about the value the @coveo/service-core can bring here. Plz poke me if there is anything specific you want us to check. :pray: 

Meanwhile I'll approve in order to not block anything.

File: packages/atomic/src/utils/init-queue.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+function getParentReadyMap(): Map<Element, boolean> {
+  return getInitQueueNamespace().parentReadyMap;
+}
+
+export function markParentAsReady(parent: Element) {
+  const parentReadyMap = getParentReadyMap();
+  parentReadyMap.set(parent, true);
+  const eventQueueMap = getEventQueueMap();
+  const eventQueue = eventQueueMap.get(parent) || [];
+  while (eventQueue.length > 0) {
```
Comment (y-lakhdar) [inline packages/atomic/src/utils/init-queue.ts:46]: This is inefficient because `Array.prototype.shift` reindexes the array on every call. Can't we use `.pop()` instead since it is an `O(1)` operation?

If the dispatch order is really important, then we should have a test for that.

File: packages/atomic/src/utils/init-queue.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    const {event, element} = eventQueue.shift()!;
+    element.dispatchEvent(event);
+  }
+  parent.dispatchEvent(new CustomEvent('atomic/parentReady', {bubbles: true}));
```
Comment (y-lakhdar) [inline packages/atomic/src/utils/init-queue.ts:50]: ### nitpick improvement 
Can clean the map by removing the parent now that all the queue has been processed
```
eventQueueMap.delete(parent);
```

File: packages/atomic/src/utils/init-queue.spec.ts [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    document.body.appendChild(parent);
+    document.body.appendChild(child);
+  });
+
+  afterEach(() => {
+    document.body.removeChild(parent);
+    document.body.removeChild(child);
+  });
+
+  it('should mark parent as ready', () => {
```
Comment (y-lakhdar) [inline packages/atomic/src/utils/init-queue.spec.ts:29]: The tests do not verify the behavior when `markParentAsReady` is called multiple times for the same parent.
```ts
it('should handle multiple calls to markParentAsReady gracefully', () => {
  markParentAsReady(parent);
  expect(isParentReady(parent)).toBe(true);
  markParentAsReady(parent); // Call again
  expect(isParentReady(parent)).toBe(true); // Should still be true
});
```
Subsequent calls should not behave differently.

File: packages/atomic/src/utils/init-queue.spec.ts [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  it('should queue event for parent', () => {
+    queueEventForParent(parent, event, child);
+    expect(isParentReady(parent)).toBe(false);
+  });
+
+  it('should dispatch queued events when parent is marked as ready', () => {
+    const spy = vi.spyOn(child, 'dispatchEvent');
+    queueEventForParent(parent, event, child);
+    markParentAsReady(parent);
+    expect(spy).toHaveBeenCalledWith(event);
```
Comment (y-lakhdar) [inline packages/atomic/src/utils/init-queue.spec.ts:43]: not sure I understand this test. Should the event be dispatched immediately?
```
expect(spy).toHaveBeenCalledWith(event); // Event should be dispatched immediately
```

File: packages/atomic/src/utils/init-queue.ts (hunk 14 above)
Comment (y-lakhdar) [reply packages/atomic/src/utils/init-queue.ts:50]: and maybe add this if we clean up the queue map
```ts
  it('should remove parent from eventQueueMap after marking as ready', () => {
    queueEventForParent(parent, event, child);
    markParentAsReady(parent);
    const {eventQueueMap} = window.initQueueNamespace;
    expect(eventQueueMap.has(parent)).toBe(false);
  });
```

**Review by y-lakhdar — APPROVED** [review]

File: packages/atomic/src/utils/init-queue.ts (hunk 13 above)
Comment (louis-bompart) [reply packages/atomic/src/utils/init-queue.ts:46]: Good point. IIRC, order matters (i.e. we need a FIFO structure, i.e. queue. Stack doesn't cut it)
So here we have a O(n^2). oops.
Assuming the queue is immutable once we start reading it, we should be able to make due with O(n) by reversing the array before reading it with pop.

Comment (louis-bompart) [reply packages/atomic/src/utils/init-queue.ts:46]: If we want to go further, we could go O(1) but we'd need to craft a LinkedList, but that's fricking overkill when we consider that this is only there till we get rid of Stencil.

File: packages/atomic/project.json [hunk 17]
```diff
@@ -54,7 +56,6 @@
       }
     },
     "release:phase1": {},
-    "release:phase3": {},
```
Comment (alexprudhomme) [inline packages/atomic/project.json:57]: @fpbrault Due to a bad merge conflict I guess you removed this which breaks the npm prerelease & release of atomic. Ill take care of it here https://github.com/coveo/ui-kit/pull/5051
